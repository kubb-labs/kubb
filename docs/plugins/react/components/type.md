---
layout: doc

title: \@kubb/react
outline: deep
---

# Type

```tsx twoslash
import React from "react"
import { createRoot, Type } from '@kubb/react'

const root = createRoot()

const Component = () => {
  return (
    <Type name="Data">
      string
    </Type>
  )
}

root.render(<Component />)
root.output
//   ^?
```


```typescript
type Data = string
```

## API

```tsx twoslash
import React from "react"
import { Type } from '@kubb/react'

type Props = React.ComponentProps<typeof Type>
```

| Property | Description                                                  | Type                     | Default |
| -------- | ------------------------------------------------------------ | ------------------------ | ------- |
| name     | Name of the type, this needs to start with a capital letter. | `string`                 | -       |
| export   | Does this type need to be exported.                          | `boolean \|  undefined`  | -       |
| JSDoc    | Options for JSdocs                                           | `JSDoc \|  undefined`    | -       |
| children |                                                              | `KubbNode \|  undefined` | -       |
