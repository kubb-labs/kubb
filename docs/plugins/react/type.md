---
layout: doc

title: \@kubb/react
outline: deep
---

# Type

::: code-group

```tsx [simple]
import { createRoot, Type } from '@kubb/react'

const root = createRoot()

const Component = () => {
  return (
    <Type name="Data">
      string
    </Type>
  )
}

root.render(<Component />)

return root.output
```

:::

::: code-group

```typescript [simple]
type Data = string
```

:::

## API

| Property | Description                                                  | Type                     | Default |
| -------- | ------------------------------------------------------------ | ------------------------ | ------- |
| name     | Name of the type, this needs to start with a capital letter. | `string`                 | -       |
| export   | Does this type need to be exported.                          | `boolean \|  undefined`  | -       |
| JSDoc    | Options for JSdocs                                           | `JSDoc \|  undefined`    | -       |
| children |                                                              | `KubbNode \|  undefined` | -       |
