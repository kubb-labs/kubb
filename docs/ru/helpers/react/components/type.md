---
layout: doc

title: \@kubb/react
outline: deep
---

# Type

```tsx
import React from "react"
import { createRoot, Type } from '@kubb/react'

const root = createRoot()

const Component = () => {
  return (
    <Type name="Data">
      string
    </Type>
  )
}

root.render(<Component />)
root.output
//   ^?
```


```typescript
type Data = string
```

## API

```tsx
import React from "react"
import { Type } from '@kubb/react'

type Props = React.ComponentProps<typeof Type>
```

| свойство | описание                                                 | тип                      | по умолчанию |
|----------|----------------------------------------------------------|--------------------------|--------------|
| name     | имя типа, должно начинаться с заглавной буквы.           | `string`                 | -            |
| export   | нужно ли экспортировать этот тип.                        | `boolean \|  undefined`  | -            |
| JSDoc    | опции для JSdocs                                         | `JSDoc \|  undefined`    | -            |
| children |                                                          | `KubbNode \|  undefined` | -            |
