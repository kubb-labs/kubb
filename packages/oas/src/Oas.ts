import BaseOas from 'oas'
import OASNormalize from 'oas-normalize'
import type { Operation } from 'oas/operation'
import type { MediaTypeObject, OASDocument, ResponseObject, SchemaObject, User } from 'oas/types'
import { findSchemaDefinition, matchesMimeType } from 'oas/utils'
import type { contentType } from './types.ts'
import { isReference } from './utils.ts'
type Options = {
  contentType?: contentType
}

export class Oas<const TOAS = unknown> extends BaseOas {
  #options: Options = {}
  document: TOAS = undefined as unknown as TOAS

  constructor({ oas, user }: { oas: TOAS | OASDocument | string; user?: User }, options: Options = {}) {
    if (typeof oas === 'string') {
      oas = JSON.parse(oas)
    }

    super(oas as OASDocument, user)

    this.document = oas as TOAS
    this.#options = options
  }

  dereferenceWithRef(schema?: unknown) {
    if (isReference(schema)) {
      return {
        ...findSchemaDefinition(schema.$ref, this.api),
        $ref: schema.$ref,
      }
    }

    return schema
  }

  /**
   * Oas does not have a getResponseBody(contentType)
   */
  #getResponseBodyFactory(responseBody: boolean | ResponseObject): (contentType?: string) => MediaTypeObject | false | [string, MediaTypeObject, ...string[]] {
    function hasResponseBody(res = responseBody): res is ResponseObject {
      return !!res
    }

    return (contentType) => {
      if (!hasResponseBody(responseBody)) {
        return false
      }

      if (isReference(responseBody)) {
        // If the request body is still a `$ref` pointer we should return false because this library
        // assumes that you've run dereferencing beforehand.
        return false
      }

      if (!responseBody.content) {
        return false
      }

      if (contentType) {
        if (!(contentType in responseBody.content)) {
          return false
        }

        return responseBody.content[contentType]!
      }

      // Since no media type was supplied we need to find either the first JSON-like media type that
      // we've got, or the first available of anything else if no JSON-like media types are present.
      let availablecontentType: string | undefined = undefined
      const contentTypes = Object.keys(responseBody.content)
      contentTypes.forEach((mt: string) => {
        if (!availablecontentType && matchesMimeType.json(mt)) {
          availablecontentType = mt
        }
      })

      if (!availablecontentType) {
        contentTypes.forEach((mt: string) => {
          if (!availablecontentType) {
            availablecontentType = mt
          }
        })
      }

      if (availablecontentType) {
        return [availablecontentType, responseBody.content[availablecontentType]!, ...(responseBody.description ? [responseBody.description] : [])]
      }

      return false
    }
  }

  getResponseSchema(operation: Operation, statusCode: string | number): SchemaObject {
    if (operation.schema.responses) {
      Object.keys(operation.schema.responses).forEach((key) => {
        const schema = operation.schema.responses![key]
        const $ref = isReference(schema) ? schema.$ref : undefined

        if (schema && $ref) {
          operation.schema.responses![key] = findSchemaDefinition($ref, this.api)
        }
      })
    }

    const getResponseBody = this.#getResponseBodyFactory(operation.getResponseByStatusCode(statusCode))

    const { contentType } = this.#options
    const responseBody = getResponseBody(contentType)

    if (responseBody === false) {
      // return empty object because response will always be defined(request does not need a body)
      return {}
    }

    const schema = Array.isArray(responseBody) ? responseBody[1].schema : responseBody.schema

    if (!schema) {
      // return empty object because response will always be defined(request does not need a body)

      return {}
    }

    return this.dereferenceWithRef(schema)
  }

  getRequestSchema(operation: Operation): SchemaObject | undefined {
    const { contentType } = this.#options

    if (operation.schema.requestBody) {
      operation.schema.requestBody = this.dereferenceWithRef(operation.schema.requestBody)
    }

    const requestBody = operation.getRequestBody(contentType)

    if (requestBody === false) {
      return undefined
    }

    const schema = Array.isArray(requestBody) ? requestBody[1].schema : requestBody.schema

    if (!schema) {
      return undefined
    }

    return this.dereferenceWithRef(schema)
  }

  getParametersSchema(operation: Operation, inKey: 'path' | 'query' | 'header'): SchemaObject | null {
    const { contentType = operation.getContentType() } = this.#options
    const params = operation
      .getParameters()
      .map((schema) => {
        return this.dereferenceWithRef(schema)
      })
      .filter((v) => v.in === inKey)

    if (!params.length) {
      return null
    }

    return params.reduce(
      (schema, pathParameters) => {
        const property = pathParameters.content?.[contentType]?.schema ?? (pathParameters.schema as SchemaObject)
        const required = [...(schema.required || ([] as any)), pathParameters.required ? pathParameters.name : undefined].filter(Boolean)

        return {
          ...schema,
          description: schema.description,
          deprecated: schema.deprecated,
          example: schema.example,
          required,
          properties: {
            ...schema.properties,
            [pathParameters.name]: {
              description: pathParameters.description,
              ...property,
            },
          },
        }
      },
      { type: 'object', required: [], properties: {} } as SchemaObject,
    )
  }

  async valdiate() {
    const oasNormalize = new OASNormalize(this.api, {
      enablePaths: true,
      colorizeErrors: true,
    })

    await oasNormalize.validate()
  }
}
