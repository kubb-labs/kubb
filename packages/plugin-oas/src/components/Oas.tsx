import { createContext } from '@kubb/react'

import { Operation } from './Operation.tsx'
import { Schema } from './Schema.tsx'

import type { Oas as OasType, Operation as OperationType } from '@kubb/oas'
import type { KubbNode } from '@kubb/react'
import type { OperationGenerator } from '../OperationGenerator.ts'

type Props = {
  oas: OasType
  operations?: OperationType[]
  /**
   * @deprecated
   */
  generator?: Omit<OperationGenerator, 'build'>
  children?: KubbNode
}

type OasContextProps = {
  oas?: OasType
  operations?: OperationType[]
  /**
   * @deprecated
   */
  generator?: Omit<OperationGenerator, 'build'>
}

const OasContext = createContext<OasContextProps>({})

export function Oas({ oas, children, operations, generator }: Props): KubbNode {
  return <OasContext.Provider value={{ oas, generator, operations }}>{children}</OasContext.Provider>
}

Oas.Context = OasContext
Oas.Operation = Operation
Oas.Schema = Schema
