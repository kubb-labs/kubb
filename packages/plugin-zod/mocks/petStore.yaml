openapi: 3.1.0
info:
  version: 1.0.0
  title: Swagger Petstore
  license:
    name: MIT
servers:
  - url: http://petstore.swagger.io/v1
paths:
  /pets:
    get:
      summary: List all pets
      operationId: listPets
      tags:
        - pets
      parameters:
        - name: limit
          in: query
          description: How many items to return at one time (max 100)
          required: false
          schema:
            type: string
        - name: offset
          in: query
          required: false
          schema:
            type: integer
            default: 0
      responses:
        '200':
          description: A paged array of pets
          headers:
            x-next:
              description: A link to the next page of responses
              schema:
                type: string
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Pets'
        default:
          description: unexpected error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Error'
    post:
      summary: Create a pet
      operationId: createPets
      tags:
        - pets
      requestBody:
        required: true
        content:
          application/json:
            schema:
              type: object
              required:
                - 'name'
                - 'tag'
              properties:
                name:
                  type: string
                tag:
                  type: string
      responses:
        '201':
          description: Null response
        default:
          description: unexpected error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Error'
  /pets/{petId}:
    get:
      summary: Info for a specific pet
      operationId: showPetById
      tags:
        - pets
      parameters:
        - name: petId
          in: path
          required: true
          description: The id of the pet to retrieve
          schema:
            type: string
        - name: testId
          in: path
          required: true
          description: The id of the pet to retrieve
          schema:
            type: string
      responses:
        '200':
          description: Expected response to a valid request
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Pet'
        default:
          description: unexpected error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Error'
components:
  schemas:
    Order:
      type: object
      properties:
        id:
          type: integer
          format: int64
          example: 10
        petId:
          type: integer
          format: int64
          example: 198772
        quantity:
          type: integer
          format: int32
          example: 7
        shipDate:
          type: string
          format: date-time
        status:
          type: string
          description: Order Status
          example: approved
          enum:
            - placed
            - approved
            - delivered
        complete:
          type: boolean
      xml:
        name: order
    Pet:
      type: [object]
      required:
        - id
        - name
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
        date:
          type: string
          format: date
        uuid:
          type: string
          format: uuid
        email:
          type: string
          format: email
        pattern:
          type: string
          pattern: "^[a-zA-Z0-9]{3}$"
        tag:
          type: string
          maxLength: 100
          minLength: 5
    PetArray:
      description: List of Pet object
      type: array
      minItems: 1
      maxItems: 3
      items:
        $ref: '#/components/schemas/Pet'
    OptionalPet:
      type: object
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
        tag:
          type: string
    PetTuple:
      maxItems: 2
      minItems: 2
      prefixItems:
        - type: string
        - type: string
      title: Attributes
      type: array
      description: Tuple of exact length 2
      PetTuple:
        maxItems: 2
        minItems: 2
        prefixItems:
          - type: string
          - type: string
        type: array
        description: Tuple of exact length 2
    PetTupleObject:
      type: object
      description: Tuple of exact length 2 nested in an object
      properties:
        tupleProperty:
          type: array,
          maxItems: 2,
          minItems: 2,
          prefixItems:
            - type: string
            - type: string
    Pets:
      type: array
      items: {
        type: 'object',
        required: ['id', 'name'],
        properties: { id: { type: 'integer', format: 'int64' }, name: { type: 'string' }, tag: { type: 'string' } },
      }
    Toy:
      type: object
      required:
        - id
      properties:
        id:
          type: string
          format: uuid
        name:
          type: string
        description:
          type: string
          default: null
          nullable: true
      default: null
      nullable: true
    Error:
      type: object
      required:
        - code
        - message
      properties:
        code:
          type: integer
          format: int32
        message:
          type: string
