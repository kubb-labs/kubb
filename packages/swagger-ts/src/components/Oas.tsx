import { File, Type, usePlugin } from '@kubb/react'
import { useFile } from '@kubb/react'
import { useOas } from '@kubb/swagger/hooks'

import type { OasTypes } from '@kubb/swagger'
import type { ReactNode } from 'react'
import type { FileMeta, PluginOptions } from '../types.ts'

type TemplateProps = {
  /**
   * Name of the function
   */
  name: string
  typeName: string
  api: OasTypes.OASDocument
}

function Template({
  name,
  typeName,
  api,
}: TemplateProps): ReactNode {
  return (
    <>
      {`export const ${name} = ${JSON.stringify(api, undefined, 2)} as const`}
      <br />
      <Type name={typeName} export>
        {`Parse<typeof ${name}>`}
      </Type>
    </>
  )
}

const defaultTemplates = { default: Template } as const

type Props = {
  name: string
  typeName: string
  /**
   * This will make it possible to override the default behaviour.
   */
  Template?: React.ComponentType<React.ComponentProps<typeof Template>>
}

export function Oas({
  name,
  typeName,
  Template = defaultTemplates.default,
}: Props): ReactNode {
  const oas = useOas()

  return <Template name={name} typeName={typeName} api={oas.api} />
}

type FileProps = {
  name: string
  typeName: string
  /**
   * This will make it possible to override the default behaviour.
   */
  templates?: typeof defaultTemplates
}

Oas.File = function({ name, typeName, templates = defaultTemplates }: FileProps): ReactNode {
  const { key: pluginKey } = usePlugin<PluginOptions>()
  const file = useFile({ name, pluginKey })

  const Template = templates.default

  return (
    <File<FileMeta>
      baseName={file.baseName}
      path={file.path}
      meta={file.meta}
    >
      <File.Import name={['Parse']} path="@kubb/swagger-ts/infer" isTypeOnly />
      <File.Source>
        <Oas Template={Template} name={name} typeName={typeName} />
      </File.Source>
    </File>
  )
}

Oas.templates = defaultTemplates
