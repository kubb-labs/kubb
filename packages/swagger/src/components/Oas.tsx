import { createContext } from '@kubb/react'

import { Operation } from './Operation.tsx'
import { Schema } from './Schema.tsx'

import type { KubbNode } from '@kubb/react'
import type { Oas as OasType, Operation as OperationType } from '../oas/index.ts'
import type { OperationSchemas } from '../types.ts'

export type GetOperationSchemas = (operation: OperationType, statusCode?: string | number) => OperationSchemas

type Props = {
  oas: OasType
  operations?: OperationType[]
  getOperationSchemas?: GetOperationSchemas
  children?: KubbNode
}

type OasContextProps = {
  oas?: OasType
  operations?: OperationType[]
  getOperationSchemas?: GetOperationSchemas
}

const OasContext = createContext<OasContextProps>({})

export function Oas({ oas, children, operations, getOperationSchemas }: Props): KubbNode {
  return <OasContext.Provider value={{ oas, getOperationSchemas, operations }}>{children}</OasContext.Provider>
}

Oas.Context = OasContext
Oas.Operation = Operation
Oas.Schema = Schema
