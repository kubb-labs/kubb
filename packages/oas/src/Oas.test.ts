import path from 'node:path'

import yaml from '@stoplight/yaml'
import { expectTypeOf } from 'expect-type'

import { petStore } from '../mocks/petStore.ts'

import { Oas } from './Oas.ts'
import type { Infer, MethodMap, Model, PathMap, RequestParams, Response } from './infer/index.ts'
import { parse } from './parser/index.ts'

describe('swagger Infer', () => {
  const oas = new Oas({ oas: petStore })
  type Oas = Infer<typeof oas.document>
  //   ^?
  type Paths = keyof PathMap<Oas>
  //    ^?
  type Methods = keyof MethodMap<Oas, '/pet'>
  //     ^?
  type UserModel = Model<Oas, 'User'>
  //     ^?
  type UserRequestParams = RequestParams<Oas, '/pet', 'post'>
  type UserResponse = Response<Oas, '/pet', 'post', '200'>
  test('types', () => {
    expectTypeOf<Paths>().not.toBeUndefined()
    expectTypeOf<Methods>().toMatchTypeOf<'post' | 'put'>()
    expectTypeOf<UserModel>().toMatchTypeOf<{
      username?: string | undefined
    }>()
    expectTypeOf<UserRequestParams['json']>().toMatchTypeOf<{
      status?: 'available' | 'pending' | 'sold' | undefined
    }>()
    expectTypeOf<UserResponse>().toMatchTypeOf<{
      status?: 'available' | 'pending' | 'sold' | undefined
    }>()
  })
})

describe('Oas filter', async () => {
  const petStorePath = path.resolve(__dirname, '../mocks/petStore.yaml')

  test('Filtering on operationId', async () => {
    const oas = await parse(petStorePath, {
      filterSet: {
        inverseMethods: ['put'],
        operationIds: ['updateUser'],
      },
    })

    expect(yaml.safeStringify(oas.api)).toMatchSnapshot()
  })
})
