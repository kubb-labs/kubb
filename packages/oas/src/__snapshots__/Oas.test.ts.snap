// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Oas filter > Filtering on operationId 1`] = `
"openapi: 3.0.3
info:
  title: Swagger Petstore - OpenAPI 3.0
  description: >-
    This is a sample Pet Store Server based on the OpenAPI 3.0 specification. 
    You can find out more about

    Swagger at [https://swagger.io](https://swagger.io). In the third iteration
    of the pet store, we've switched to the design first approach!

    You can now help us improve the API whether it's by making changes to the
    definition itself or to the code.

    That way, with time, we can improve the API in general, and expose some of
    the new features in OAS3.


    Some useful links:

    - [The Pet Store
    repository](https://github.com/swagger-api/swagger-petstore)

    - [The source API definition for the Pet
    Store](https://github.com/swagger-api/swagger-petstore/blob/master/src/main/resources/openapi.yaml)
  termsOfService: 'http://swagger.io/terms/'
  contact:
    email: apiteam@swagger.io
  license:
    name: Apache 2.0
    url: 'http://www.apache.org/licenses/LICENSE-2.0.html'
  version: 1.0.11
servers:
  - url: 'https://petstore3.swagger.io/api/v3'
paths:
  /pet:
    put:
      operationId: updatePet
      summary: Update an existing pet
      description: Update an existing pet by Id
      requestBody:
        description: Update an existent pet in the store
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Pet'
          application/x-www-form-urlencoded:
            schema:
              $ref: '#/components/schemas/Pet'
          application/xml:
            schema:
              $ref: '#/components/schemas/Pet'
      responses:
        '200':
          description: Successful operation
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Pet'
            application/xml:
              schema:
                $ref: '#/components/schemas/Pet'
        '400':
          description: Invalid ID supplied
        '404':
          description: Pet not found
        '405':
          description: Validation exception
      security:
        - petstore_auth:
            - 'write:pets'
            - 'read:pets'
      tags:
        - pet
components:
  schemas:
    Category:
      type: object
      properties:
        id:
          type: integer
          format: int64
          example: 1
        name:
          type: string
          example: Dogs
      xml:
        name: category
    tag.Tag:
      type: object
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
      xml:
        name: tag
    Pet:
      type: object
      properties:
        id:
          type: integer
          format: int64
          example: 10
          readOnly: true
        name:
          type: string
          example: doggie
          writeOnly: true
        category:
          $ref: '#/components/schemas/Category'
        photoUrls:
          type: array
          items:
            type: string
            xml:
              name: photoUrl
          xml:
            wrapped: true
        tags:
          type: array
          items:
            $ref: '#/components/schemas/tag.Tag'
          xml:
            wrapped: true
        status:
          description: pet status in the store
          type: string
          enum:
            - available
            - pending
            - sold
      required:
        - name
        - photoUrls
      xml:
        name: pet
  securitySchemes:
    petstore_auth:
      type: oauth2
      flows:
        implicit:
          authorizationUrl: 'https://petstore3.swagger.io/oauth/authorize'
          scopes:
            'write:pets': modify pets in your account
            'read:pets': read your pets
    api_key:
      type: apiKey
      name: api_key
      in: header
tags:
  - name: pet
    description: Everything about your Pets
    externalDocs:
      description: Find out more
      url: 'http://swagger.io'
  - name: store
    description: Access to Petstore orders
    externalDocs:
      description: Find out more about our store
      url: 'http://swagger.io'
  - name: user
    description: Operations about user
externalDocs:
  description: Find out more about Swagger
  url: 'http://swagger.io'
"
`;
